# scVMAP database

> scVMAP front-end: https://github.com/YuZhengM/scvmap_web

> scVMAP back-end: https://github.com/YuZhengM/scvmap

## 1. Path storage instructions

> Path storage information during data processing.

> `path` is the root path

- 📁`{path}`
  - 📁database: Store `SCVdb` database data
    - 📁code: Store the code for running database data
      - 📁`{server}`: Represent different servers and store the code for executing g-chromVAR and SCAVENGE methods
        - 📄library.R: Referenced R packages
        - 📄static_function.R: Processing of single-cell and integration code
        - 📄integration.R: Run data for multiple traits or diseases in a loop
        - 📄run.R: The main execution code
    - 📁sc_variant: Store the root path of database data output
      - 📁result: Store the results of g-chromVAR and SCAVENGE methods
        - 📁`{scATAC-seq}`: Represent different single-cell data and store the result data of integrating all traits or diseases for a certain single-cell data
          - 📄`{scATAC-seq}__{genome}__{trait_label_file}__mat.txt`: Store information files such as TRS and enrichment status
          - 📄`{scATAC-seq}__{genome}__{trait_label_file}__mat_info.rda`: Store the result txt and mutual-KNN information files
      - 📁scATAC: Store the pre-processing files and post-processing result files of single-cell data
        - 📁`{scATAC-seq}`: Represent different single-cell data and store the pre-processing files and post-processing result files of a certain single-cell data
          - 📄`{scATAC-seq}.txt`: Cell annotation file
          - 📄`{scATAC-seq}_all.rda`: Single-cell processing file, which is read when integrating with traits or diseases
          - 📄`{scATAC-seq}_SE_gvar_SE_gvar_bg.rda`: Intermediate file for single-cell processing, with less content than `{scATAC-seq}_all.rda`
          - 📄`{scATAC-seq}_trs_scavenge_data.h5ad`: The result data of integrating all traits or diseases for a certain single-cell data in h5ad file format
      - 📁table: Store data related to database construction
        - 📁download: Store scATAC-seq data processed by SnapATAC2
          - 📁scatac: Store scATAC-seq data processed by SnapATAC2
            - 📄`{scATAC-seq}_snapATAC2.h5ad`: Data downloaded from the download page on the website
          - 📄cp.sh: Copy all scATAC-seq data processed by SnapATAC2 to the scatac directory in the same path
        - 📁homer: Store the transcription factor data of traits or diseases of interest generated by HOMER
          - 📁hg19: Store the transcription factor data of traits or diseases of interest with the reference genome hg19
            - 📄`t_homer_{group}`.txt: Corresponds to the `t_homer_hg19_{group}` table in the database
          - 📁hg38: Store the transcription factor data of traits or diseases of interest with the reference genome hg38
            - 📄`t_homer_{group}`.txt: Corresponds to the `t_homer_hg38_{group}` table in the database
          - 📄t_homer_tf_trait_count.txt: Store the number of TFs in trait or disease samples and transcription factors under different reference genomes, corresponding to the `t_tf_trait_count` table in the database
          - 📄t_homer_trait_tf.txt: Store the mapping relationship between trait or disease samples and transcription factors under different reference genomes
          - 📄t_homer_trait_tf_hg19.txt: Store the mapping relationship between hg19 trait or disease samples and transcription factors, corresponding to the `t_trait_tf_hg19` table in the database
          - 📄t_homer_trait_tf_hg38.txt: Store the mapping relationship between hg38 trait or disease samples and transcription factors, corresponding to the `t_trait_tf_hg38` table in the database
          - 📄trait_tf_hg19.txt: This file combines the data of `hg19` files (not used, kept)
          - 📄trait_tf_hg38.txt: This file combines the data of `hg38` files (not used, kept)
        - 📁magma: Store the gene data of traits or diseases of interest generated by MAGMA
          - 📁gene_enrichment_trait: Store the GO, KEGG, and other gene enrichment functions of genes of interest in traits or diseases
            - 📁hg19: Store the GO, KEGG, and other gene enrichment functions of genes of interest in traits or diseases with the reference genome hg19
            - 📁hg38: Store the GO, KEGG, and other gene enrichment functions of genes of interest in traits or diseases with the reference genome hg38
          - 📁gene_enrichment_trait_table: Store the GO, KEGG, and other gene enrichment functions of genes of interest in traits or diseases
            - 📁hg19: Store the GO, KEGG, and other gene enrichment functions of genes of interest in traits or diseases with the reference genome hg19
              - 📄`t_gene_enrichment_trait_hg19_{group}.txt`: Corresponds to the `t_gene_enrichment_trait_hg19_{group}` table in the database
            - 📁hg38: Store the GO, KEGG, and other gene enrichment functions of genes of interest in traits or diseases with the reference genome hg38
              - 📄`t_gene_enrichment_trait_hg38_{group}.txt`: Corresponds to the `t_gene_enrichment_trait_hg38_{group}` table in the database
          - 📁hg19: Store the gene data of traits or diseases of interest with the reference genome hg19 (with enrichment scores)
            - 📄`t_magma_{group}.txt`: Corresponds to the `t_magma_hg19_{group}` table in the database
          - 📁hg19_anno: Store the gene data of traits or diseases of interest with the reference genome hg19 (with mutation and gene mapping)
            - 📄`t_magma_{group}.txt`: Corresponds to the `t_magma_anno_hg19_{group}` table in the database
          - 📁hg38: Store the gene data of traits or diseases of interest with the reference genome hg38 (with enrichment scores)
            - 📄`t_magma_{group}.txt`: Corresponds to the `t_magma_hg38_{group}` table in the database
          - 📁hg38_anno: Store the gene data of traits or diseases of interest with the reference genome hg38 (with mutation and gene mapping)
            - 📄`t_magma_{group}.txt`: Corresponds to the `t_magma_anno_hg38_{group}` table in the database
          - 📄t_magma_gene_trait_count.txt: Store the number of Genes in trait or disease samples and transcription factors under different reference genomes, corresponding to the `t_gene_trait_count` table in the database
          - 📄t_magma.txt: Store the mapping relationship between trait or disease samples and genes under different reference genomes
          - 📄t_magma_hg19.txt: Store the mapping relationship between hg19 trait or disease samples and genes, corresponding to the `t_trait_gene_hg19` table in the database
          - 📄t_magma_hg38.txt: Store the mapping relationship between hg38 trait or disease samples and genes, corresponding to the `t_trait_gene_hg38` table in the database
        - 📁scatac: Store the database content related to scATAC-seq data
          - 📁difference_gene: Differential gene enrichment of single-cell samples
            - `{scATAC-seq}_difference_gene_data.txt`: Differential gene enrichment of a certain single-cell sample (not used, kept)
          - 📁difference_tf: Differential transcription factor enrichment of single-cell samples
            - `{scATAC-seq}_difference_tf_data.txt`: Differential transcription factor enrichment of a certain single-cell sample (not used, kept)
          - 📁gene_enrichment_table: Differential gene GO, KEGG, and other pathway enrichment of single-cell samples
            - `{scATAC-seq}_gene_enrichment_data.txt`: Differential gene GO, KEGG, and other pathway enrichment of a certain single-cell sample, corresponding to the `t_gene_enrichment_{scATAC-seq}` table in the database
          - 📄difference_gene_data.txt: Differential gene enrichment of all single-cell samples, corresponding to the `t_difference_gene` table in the database
          - 📄difference_tf_data.txt: Differential transcription factor enrichment of all single-cell samples, corresponding to the `t_difference_tf` table in the database
          - 📄gene_enrichment_table_data.txt: Differential gene GO, KEGG, and other pathway enrichment of all single-cell samples (not used, kept)
          - 📄sample_gene_data.txt: File storing the mapping relationship between single-cell samples and genes, corresponding to the `t_sample_gene` table in the database
        - 📁trait_variant_overlap: Store the overlap status between all single-cell data and all traits or diseases (overlap status: !whether all TRS scores are zero)
          - 📄gchromvar_sample_enrichment.txt: Store the overlap status of the g-chromVAR method in a txt file, corresponding to the `t_difference_tf` table in the database
          - 📄scavenge_sample_enrichment.txt: Store the overlap status of the SCAVENGE method in a txt file, corresponding to the `t_difference_tf` table in the database
          - 📄trs_overlap_gchromvar.h5ad: Store the overlap status of the g-chromVAR method in an h5ad file (not used, kept)
          - 📄trs_overlap_scavenge.h5ad: Store the overlap status of the SCAVENGE method in an h5ad file (not used, kept)
        - 📁trs: Store the data content for visual display when the website reads files (the result h5ad files are split into 100 sub - files here)
          - 📁`{scATAC-seq}`: Represent different single-cell data and store the results of a certain single-cell data integrated with grouped traits or diseases
            - 📁cell_type: TRS results with cell type as the resolution
              - 📄`{scATAC-seq}_cell_type_trs_gchromvar.h5ad`: Cell types-traits matrix calculated by the g-chromVAR method
              - 📄`{scATAC-seq}_cell_type_trs_scavenge.h5ad`: Cell types-traits matrix calculated by the SCAVENGE method
            - 📁chunk: TRS results
              - 📄`{scATAC-seq}_{group}_trs_gchromvar.h5ad`: Cells-traits matrix calculated by the g-chromVAR method
              - 📄`{scATAC-seq}_{group}_trs_scavenge.h5ad`: Cells-traits matrix calculated by the SCAVENGE method
        - 📁trs_big: Store the files downloaded from the website (here store the result h5ad files without splitting)
          - 📁`{scATAC-seq}`: Represent different single-cell data and store the results of a certain single-cell data integrated with all traits or diseases
              - 📄`{scATAC-seq}_trs_gchromvar.h5ad`: Cells-traits matrix calculated by the g-chromVAR method
              - 📄`{scATAC-seq}_trs_scavenge.h5ad`: Cells-traits matrix calculated by the SCAVENGE method
        - 📁variant: Store the data of the website's mutation information database
          - 📁hg19: Store the fine-mapping result data with the reference genome hg19
            - 📄`t_variant_{group}_hg19.txt`: Corresponds to the `t_variant_hg19_{group}` table in the database
          - 📁hg38: Store the fine-mapping result data with the reference genome hg38
            - 📄`t_variant_{group}_hg38.txt`: Corresponds to the `t_variant_hg38_{group}` table in the database
          - 📄t_trait_chr_count.txt: Store the number of different chromatin in traits or diseases, corresponding to the `t_trait_chr_count` table in the database
      - 📁variant: Store the input fine-mapping result data for running data
        - 📁hg19: Store the fine-mapping result data with the reference genome hg19
        - 📁hg38: Store the fine-mapping result data with the reference genome hg38
        - 📄fine_mapping_result.pkl: Store the fine-mapping result data of all reference genomes in a file
        - 📄fine_mapping_result_hg19.pkl: Store the fine-mapping result data with the reference genome hg19 in a file
        - 📄fine_mapping_result_hg38.pkl: Store the fine-mapping result data with the reference genome hg38 in a file
  - 📁gene: Store gene-related data
    - 📁download: Store the gene version consistent with SnapATAC2 currently
      - 📄gencode.v41.annotation.gtf.gz: Reference genome hg38
      - 📄gencode.v41lift37.annotation.gtf.gz: Reference genome hg19
    - 📁result: Store the standardized tabular data processed from the downloaded data
      - 📄gene_hg19_all.txt: Tabular data containing all columns
      - 📄gene_hg38_all.txt: Tabular data containing all columns
      - 📄gene_data.txt: Data input into the database, corresponding to the `t_gene` table in the database
      - 📄gene_hg19_data.txt: Extract hg19 data from the gene_data.txt table
      - 📄gene_hg38_data.txt: Extract hg38 data from the gene_data.txt table
    - 📁annotation: Store files related to gene annotation
      - 📁dbSNP: Files downloaded from dbSNP
        - 📁common_snp_chunk: Store the chunked data of common SNPs split by reference genome
          - 📁hg19: Store the chunked data of common SNPs with reference genome hg19
          - 📁hg38: Store the chunked data of common SNPs with reference genome hg38
        - 📄dbsnp_common_snp_hg19.txt: Store the text-format data of common dbSNP SNPs with reference genome hg19
        - 📄dbsnp_common_snp_hg19.cvf: Downloaded common SNP data for hg19
        - 📄dbsnp_common_snp_hg38.txt: Store the text-format data of common dbSNP SNPs with reference genome hg38
        - 📄dbsnp_common_snp_hg38.cvf: Downloaded common SNP data for hg38
      - 📁dbSUPER: Files downloaded from dbSUPER
        - 📁all_hg19_bed: Downloaded SE data for hg19
        - 📁liftOver: Store files for reference genome conversion
          - 📁input: Store original files before conversion
          - 📁output: Store converted result files
          - 📁result: Store finalized and standardized converted files
        - 📄all_hg19_bed.zip: Store the compressed SE data for the hg19 reference genome downloaded from dbSUPER
        - 📄dbsuper_super_enhancer_hg19.txt: Store the text data of dbSUPER super-enhancers for the hg19 reference genome
        - 📄dbsuper_super_enhancer_hg38.txt: Store the text data of dbSUPER super-enhancers for the hg38 reference genome
      - 📁GTEx: Files downloaded from GTEx
        - 📁eqtl_chunk: Store the chunked eQTL data split by reference genome
          - 📁hg19: Store the chunked eQTL data with reference genome hg19
          - 📁hg38: Store the chunked eQTL data with reference genome hg38
        - 📁GTEx_Analysis_v10_eQTL_updated: Downloaded eQTL data
        - 📁liftOver: Store files for reference genome conversion
          - 📁input: Input files for conversion to the corresponding reference genome
          - 📁output: Output files after conversion to the corresponding reference genome
          - 📁result: Final gene annotation files corresponding to the reference genome
        - 📄gtex_v10_eqtl_hg19.txt: Store the GTEx v10 eQTL data with the reference genome hg19
        - 📄gtex_v10_eqtl_hg38.txt: Store the GTEx v10 eQTL data with the reference genome hg38
        - 📄gtex_v10_eqtl_hg38.tar: Downloaded GTEx v10 eQTL data in compressed format for the hg38 reference genome
      - 📁gwasATLAS: Files downloaded from gwasATLAS
        - 📁liftOver:
          - 📁input: Input files for conversion to the corresponding reference genome
          - 📁output: Output files after conversion to the corresponding reference genome
          - 📁result: Final gene annotation files corresponding to the reference genome
        - 📄gwasATLAS_v20191115.txt: Downloaded risk SNP overview data
        - 📄gwasATLAS_v20191115_riskloci.txt: Downloaded risk SNP data
        - 📄gwasatlas_v20191115_risk_snp_hg19.txt: Risk SNP data for the hg19 reference genome processed from the downloaded files
        - 📄gwasatlas_v20191115_risk_snp_hg38.txt: Risk SNP data for the hg38 reference genome processed from the downloaded files
      - 📁SEA: Files downloaded from SEA
        - 📁liftOver:
          - 📁input: Input files for conversion to the corresponding reference genome
          - 📁output: Output files after conversion to the corresponding reference genome
          - 📁result: Final gene annotation files corresponding to the reference genome
        - 📄sea_v3_enhancer_hg19.txt: Store the SEA v3 enhancer data with the reference genome hg19
        - 📄sea_v3_enhancer_hg38.txt: Store the SEA v3 enhancer data with the reference genome hg38
        - 📄sea_v3_super_enhancer_hg19.txt: Store the SEA v3 super-enhancer data with the reference genome hg19
        - 📄sea_v3_super_enhancer_hg38.txt: Store the SEA v3 super-enhancer data with the reference genome hg38
        - 📄sea_v3_super_enhancer_hg38.bed: Store the SEA v3 super-enhancer data with the reference genome hg38 in BED format
      - 📁SEdb: Files downloaded from SEdb
        - 📁enhancer_chunk: Store the chunked enhancer data split by reference genome
          - 📁hg19: Store the chunked enhancer data with reference genome hg19
          - 📁hg38: Store the chunked enhancer data with reference genome hg38
        - 📁liftOver:
          - 📁input: Input files for conversion to the corresponding reference genome
          - 📁output: Output files after conversion to the corresponding reference genome
          - 📁result: Final gene annotation files corresponding to the reference genome
        - 📄SEdb_2.0_sample_information.txt: Download the sample information of SEdb v2.0
        - 📄sedb_v2_enhancer_hg19.txt: Store the enhancer data of SEdb v2.0 with the reference genome hg19
        - 📄sedb_v2_enhancer_hg38.txt: Store the enhancer data of SEdb v2.0 with the reference genome hg38
        - 📄sedb_v2_enhancer_hg38_middle.txt: Store the intermediate enhancer data of SEdb v2.0 with the reference genome hg38 during processing
        - 📄sedb_v2_super_enhancer_hg19.txt: Store the super-enhancer data of SEdb v2.0 with the reference genome hg19
        - 📄sedb_v2_super_enhancer_hg38.txt: Store the super-enhancer data of SEdb v2.0 with the reference genome hg38
        - 📄sedb_v2_super_enhancer_hg38_middle.txt: Store the intermediate super-enhancer data of SEdb v2.0 with the reference genome hg38 during processing
        - 📄SE_package_hg38.bed: Download the super-enhancer package data in BED format with the reference genome hg38
        - 📄TE_package_hg38.bed: Download the transposable element package data in BED format with the reference genome hg38
  - 📁project_code: Store code
    - 📁scvdb_reproducibility: This folder represents the root path of this project
  - 📁scATAC: 
    - 📁`{GSE_ID}`: Store the overall processing flow data of different single-cell samples
      - 📁data: Store the data processed from the source rds files into a unified format
        - `{scATAC-seq}`: Store the data of different single-cell samples processed into a unified format
          - 📁meta: Store the metadata of single-cell samples
            - 📄barcodes.tsv: Barcodes information
            - 📄matrix.mtx: Sparse matrix information
            - 📄peaks.bed: Peak information
          - 📄annotation.txt: File of all cell annotations before processing
          - 📄annotation_stdn.txt: Unified standard file for single-cell annotations before processing
          - 📄`{scATAC-seq}_sc_atac.h5ad`: Unpreprocessed single-cell sample file
          - 📄`{scATAC-seq}_fragments.tsv.gz`: Fragments file of single-cell samples
          - 📄`{scATAC-seq}_sc_atac_snapATAC2.h5ad`: Preprocessed single-cell sample file
          - 📄`{scATAC-seq}_cell_anno.txt`: File of cell annotations after single-cell preprocessing
          - 📄`{scATAC-seq}_cell_anno_stdn.txt`: Final unified standard file for cell annotations after single-cell preprocessing
          - 📄`{scATAC-seq}_cell_type_stdn.txt`: Unified standard file for cell type annotations after single-cell preprocessing
          - 📄`{scATAC-seq}_gene_expression_data.h5ad`: Gene expression data of single-cell samples
          - 📄`{scATAC-seq}_difference_gene.h5ad`: Differential gene data of single-cell samples
          - 📄`{scATAC-seq}_gene_enrichment.txt`: Differential gene enrichment file of single-cell samples
          - 📄`{scATAC-seq}_difference_peak.h5ad`: Differential peak data of single-cell samples
          - 📄`{scATAC-seq}_tf_activity_data.h5ad`: Differential transcription factor file of single-cell samples
      - 📄`{scATAC-seq}_ATAC.rds`: Source rds files of different single-cell samples
      - 📄`{scATAC-seq}_metadata.txt`: Source annotation files of different single-cell samples
  - 📁topic: Root path information content of the SCIV algorithm
  - 📁variant: Store all fine-mapping result files (from source to final)
    - 📁source: Source data of all fine-mapping result files
      - 📁BBJ: Source data of all fine-mapping result files from the BBJ cohort
        - 📁decompression: Compress all downloaded files
          - 📄`BBJ.{trait}.Kanai2021.FINEMAP.tsv.gz`: Results generated by the FINEMAP software
          - 📄`BBJ.{trait}.Kanai2021.SuSiE.tsv.gz`: Results generated by the SuSiE software (not used)
        - 📁download: Download all fine-mapping result files of BBJ
          - 📄`hum0197.v5.finemap.{trait}.v1.zip`: Downloaded BBJ trait or disease files
        - 📁finemap: Compress all results generated by the FINEMAP software
          - 📄`BBJ.{trait}.Kanai2021.FINEMAP.tsv.txt`: Uncompressed files of the results obtained by the FINEMAP method
      - 📁CAUSALdb: Source data of all fine-mapping result files from the CAUSALdb cohort
        - 📁download: Download fine-mapping result files
          - 📁credible_set.v2.1.20240623: Folder after decompressing the fine-mapping results
            - 📁v2.1: Folder after decompressing the fine-mapping results
              - 📄credible_set.txt: Detailed content information of each trait or disease
              - 📄meta.txt: Overview file of each trait or disease
          - 📄credible_set.v2.1.20240623.tar.gz: Compressed file of the downloaded fine-mapping results
      - 📁FinnGen: Source data of all fine-mapping result files from the FinnGen cohort
        - 📁decompression: Compress all downloaded files
          - 📄`finngen_R11_{trait}.txt`: Uncompressed file
        - 📁download: Download all fine-mapping result files of FinnGen
          - 📄`finngen_R11_{trait}.gz`: Downloaded compressed file
      - 📁UKBB: Source data of all fine-mapping result files from the UKBB cohort
        - 📁decompression: Compress the UKBB_94traits_release1.bed.gz file
          - 📄UKBB_94traits_release1.bed: Uncompressed file
        - 📁download: Download the fine-mapping result files of UKBB
          - 📁release1.1: Decompress the downloaded fine-mapping result files
            - 📄README.html
            - 📄README.md
            - 📄UKBB_94traits_release1.bed.gz
            - 📄UKBB_94traits_release1.bed.gz.tbi
            - 📄UKBB_94traits_release1.cols
            - 📄UKBB_94traits_release1.traits
            - 📄UKBB_94traits_release1_regions.bed.gz
            - 📄UKBB_94traits_release1_regions.bed.gz.tbi
            - 📄UKBB_94traits_release1_regions.cols
          - 📄UKBB_94traits_release1.1.tar.gz: Downloaded fine-mapping result file
    - 📁output: Store the path of files processed into a unified standard format after decompressing all downloaded files
      - 📁`{source}`: Different source cohorts
        - 📁finemap: Form the input format for the SCAVENGE algorithm, as well as the input format for running HOMER and MAGMA
        - 📁trait: Form files with detailed content of traits or diseases
    - 📁filter: Store the fine-mapping data filtered by certain rules and converted to the corresponding reference genome data through liftOver
      - 📁finemap: Filtered finemap data in the output folder
      - 📁input_hg19: Input trait or disease data that needs to be converted to hg38
      - 📁input_hg38: Input trait or disease data that needs to be converted to hg19
      - 📁output_hg19: Output data that is already hg19 and converted to hg19
      - 📁output_hg38: Output data that is already hg38 and converted to hg38
      - 📁result_hg19: Output data of hg38 converted to a unified standard format data
      - 📁result_hg38: Output data of hg19 converted to a unified standard format data
      - 📁trait: Filtered trait data in the output folder
      - 📁unmap_hg19: Unmapped data when converting to hg19
      - 📁unmap_hg38: Unmapped data when converting to hg38
    - 📁finish: Final files after unified processing of fine-mapping result data, which are used as input files for running methods such as g-chromVAR, SCAVENGE, HOMER, and MAGMA
      - 📁hg19: Final fine-mapping result data with the reference genome hg19
        - 📄`{trait_label}.bed`: Input data for trait or disease algorithms
      - 📁hg38: Final fine-mapping result data with the reference genome hg38
        - 📄`{trait_label}.bed`: Input data for trait or disease algorithms
      - 📁trait: Final files after unified processing of fine-mapping result data, which contain detailed content information of traits or diseases in this path
        - 📄`{trait_label}.txt`: Detailed content information of traits or diseases
      - 📄variant_id.txt: File storing the mapping relationship between mutation sites and rsId
      - 📄variant_id_tmp.txt: File storing the temporary mapping relationship between mutation sites and rsId
    - 📁homer: Result data of running HOMER for all traits or diseases
      - 📁hg19: Result data of running HOMER for all traits or diseases with the reference genome hg19
        - 📁`{trait_label}`: Result data of running HOMER for this trait or disease with the reference genome hg19
      - 📁hg38: Result data of running HOMER for all traits or diseases with the reference genome hg38
        - 📁`{trait_label}`: Result data of running HOMER for this trait or disease with the reference genome hg38
    - 📁magma: Result data of running MAGMA for all traits or diseases
      - 📁gene: MAGMA reference gene data
        - 📄gene_hg19.bed
        - 📄gene_hg38.bed
      - 📁magma_input: File data information of the formed MAGMA input
        - 📁g1000_afr: African reference background template data
        - 📁g1000_amr: American reference background template data
        - 📁g1000_eas: East Asian reference background template data
        - 📁g1000_eur: European reference background template data
        - 📁g1000_sas: South Asian reference background template data
        - 📁hg19: MAGMA input files with the reference genome hg19
          - `{trait_label}.bim`: Used for the first-step gene matching SNP sites
          - `{trait_label}.txt`: Used for the second-step gene enrichment annotation
        - 📁hg38: MAGMA input files with the reference genome hg38
          - `{trait_label}.bim`: Used for the first-step gene matching SNP sites
          - `{trait_label}.txt`: Used for the second-step gene enrichment annotation
      - 📁magma_output: 
        - 📁hg19_anno: First-step output files of MAGMA with the reference genome hg19
        - 📁hg19_gene: Second-step output files of MAGMA with the reference genome hg19
        - 📁hg38_anno: First-step output files of MAGMA with the reference genome hg38
        - 📁hg38_gene: Second-step output files of MAGMA with the reference genome hg38

## 2. Database data transmission

```shell
source_path="/public/home/lcq/rgzn/yuzhengmin/keti"
target_path="/mnt/data3/workspace/SCVdb"
username="${username}"
ip="${ip}"

# test
scp -r "$source_path/database/sc_variant/table/test.txt" "$username@$ip:$target_path/data/data/download/"

# scATAC-seq
scp -r "$source_path/database/sc_variant/table/download/scatac" "$username@$ip:$target_path/data/data/download/"

# HOMER
scp -r "$source_path/database/sc_variant/table/homer/hg19" "$username@$ip:$target_path/mysql/mysqlfile/magma_homer/homer/"
scp -r "$source_path/database/sc_variant/table/homer/hg38" "$username@$ip:$target_path/mysql/mysqlfile/magma_homer/homer/"
scp -r "$source_path/database/sc_variant/table/homer/t_homer_trait_tf_hg19.txt" "$username@$ip:$target_path/mysql/mysqlfile/magma_homer/homer/"
scp -r "$source_path/database/sc_variant/table/homer/t_homer_trait_tf_hg38.txt" "$username@$ip:$target_path/mysql/mysqlfile/magma_homer/homer/"
scp -r "$source_path/database/sc_variant/table/homer/t_homer_tf_trait_count.txt" "$username@$ip:$target_path/mysql/mysqlfile/magma_homer/homer/"
scp -r "$source_path/database/sc_variant/table/homer/trait_tf" "$username@$ip:$target_path/mysql/mysqlfile/"

# MAGMA
scp -r "$source_path/database/sc_variant/table/magma/hg19" "$username@$ip:$target_path/mysql/mysqlfile/magma_homer/magma/"
scp -r "$source_path/database/sc_variant/table/magma/hg19_anno" "$username@$ip:$target_path/mysql/mysqlfile/magma_homer/magma/"
scp -r "$source_path/database/sc_variant/table/magma/hg38" "$username@$ip:$target_path/mysql/mysqlfile/magma_homer/magma/"
scp -r "$source_path/database/sc_variant/table/magma/hg38_anno" "$username@$ip:$target_path/mysql/mysqlfile/magma_homer/magma/"
scp -r "$source_path/database/sc_variant/table/magma/gene_enrichment_trait_table" "$username@$ip:$target_path/mysql/mysqlfile/magma_homer/magma/"
scp -r "$source_path/database/sc_variant/table/magma/t_magma_hg19.txt" "$username@$ip:$target_path/mysql/mysqlfile/magma_homer/magma/"
scp -r "$source_path/database/sc_variant/table/magma/t_magma_hg38.txt" "$username@$ip:$target_path/mysql/mysqlfile/magma_homer/magma/"
scp -r "$source_path/database/sc_variant/table/magma/t_magma_gene_trait_count.txt" "$username@$ip:$target_path/mysql/mysqlfile/magma_homer/magma/"
scp -r "$source_path/database/sc_variant/table/magma/trait_gene" "$username@$ip:$target_path/mysql/mysqlfile/"

scp -r "$source_path/database/sc_variant/table/trait_variant_overlap/sample" "$username@$ip:$target_path/mysql/mysqlfile/trait_sample/"
scp -r "$source_path/database/sc_variant/table/trait_variant_overlap/trait" "$username@$ip:$target_path/mysql/mysqlfile/trait_sample"

# scATAC-seq
scp -r "$source_path/database/sc_variant/table/scatac/difference_gene" "$username@$ip:$target_path/mysql/mysqlfile/scatac/"
scp -r "$source_path/database/sc_variant/table/scatac/difference_tf" "$username@$ip:$target_path/mysql/mysqlfile/scatac/"
scp -r "$source_path/database/sc_variant/table/scatac/gene_enrichment_table" "$username@$ip:$target_path/mysql/mysqlfile/scatac/"
scp -r "$source_path/database/sc_variant/table/scatac/difference_gene_data.txt" "$username@$ip:$target_path/mysql/mysqlfile/scatac/"
scp -r "$source_path/database/sc_variant/table/scatac/difference_tf_data.txt" "$username@$ip:$target_path/mysql/mysqlfile/scatac/"
scp -r "$source_path/database/sc_variant/table/scatac/sample_gene_data.txt" "$username@$ip:$target_path/mysql/mysqlfile/scatac/"

# Overlap
scp -r "$source_path/database/sc_variant/table/trait_variant_overlap/gchromvar_sample_enrichment.txt" "$username@$ip:$target_path/mysql/mysqlfile/trait_sample/"
scp -r "$source_path/database/sc_variant/table/trait_variant_overlap/scavenge_sample_enrichment.txt" "$username@$ip:$target_path/mysql/mysqlfile/trait_sample/"
scp -r "$source_path/database/sc_variant/table/trait_variant_overlap/trs_overlap_gchromvar.h5ad" "$username@$ip:$target_path/data/data/trait_variant_overlap/"
scp -r "$source_path/database/sc_variant/table/trait_variant_overlap/trs_overlap_scavenge.h5ad" "$username@$ip:$target_path/data/data/trait_variant_overlap/"

# variant
scp -r "$source_path/database/sc_variant/table/variant" "$username@$ip:$target_path/mysql/mysqlfile/"
scp -r "$source_path/variant/finish/hg19" "$username@$ip:$target_path/data/data/download/variant/"
scp -r "$source_path/variant/finish/hg19" "$username@$ip:$target_path/data/data/download/variant/"
scp -r "$source_path/variant/finish/hg19" "$username@$ip:$target_path/data/data/download/variant/"
scp -r "$source_path/variant/finish_susie/hg19" "$username@$ip:$target_path/data/data/download/variant_susie/"
scp -r "$source_path/variant/finish_susie/hg38" "$username@$ip:$target_path/data/data/download/variant_susie/"
scp -r "$source_path/variant/finish_susie/trait" "$username@$ip:$target_path/data/data/download/variant_susie/"

# TRS
scp -r "$source_path/database/sc_variant/table/trs" "$username@$ip:$target_path/data/data/"
scp -r "$source_path/database/sc_variant/table/trs_big" "$username@$ip:$target_path/data/data/download/"

# HOMER
scp -r "$source_path/variant/homer" "$username@$ip:$target_path/data/data/"
scp -r "$source_path/database/sc_variant/table/homer/trait_tf_hg19.txt" "$username@$ip:$target_path/data/data/download/magma_homer/"
scp -r "$source_path/database/sc_variant/table/homer/trait_tf_hg38.txt" "$username@$ip:$target_path/data/data/download/magma_homer/"

# MAGMA
scp -r $source_path/variant/magma/magma_output/hg19_anno/*.genes.annot $username@$ip:"$target_path/data/data/magma/magma_output/hg19_anno/"
scp -r $source_path/variant/magma/magma_output/hg19_gene/*.genes.out $username@$ip:"$target_path/data/data/magma/magma_output/hg19_gene/"
scp -r $source_path/variant/magma/magma_output/hg38_anno/*.genes.annot $username@$ip:"$target_path/data/data/magma/magma_output/hg38_anno/"
scp -r $source_path/variant/magma/magma_output/hg38_gene/*.genes.out $username@$ip:"$target_path/data/data/magma/magma_output/hg38_gene/"

scp -r "$source_path/database/sc_variant/table/magma/magma_anno_hg19_data.tar.gz" "$username@$ip:$target_path/data/data/download/magma_homer/"
scp -r "$source_path/database/sc_variant/table/magma/magma_anno_hg38_data.tar.gz" "$username@$ip:$target_path/data/data/download/magma_homer/"
scp -r "$source_path/database/sc_variant/table/magma/magma_hg19_data.tar.gz" "$username@$ip:$target_path/data/data/download/magma_homer/"
scp -r "$source_path/database/sc_variant/table/magma/magma_hg38_data.tar.gz" "$username@$ip:$target_path/data/data/download/magma_homer/"

# Difference Gene/TF
scp -r "$source_path/database/sc_variant/table/scatac/difference_gene_data.txt" "$username@$ip:$target_path/data/data/download/difference/"
scp -r "$source_path/database/sc_variant/table/scatac/difference_tf_data.txt" "$username@$ip:$target_path/data/data/download/difference/"
scp -r "$source_path/database/sc_variant/table/scatac/difference_gene" "$username@$ip:$target_path/data/data/download/difference/"
scp -r "$source_path/database/sc_variant/table/scatac/difference_gene" "$username@$ip:$target_path/mysql/mysqlfile/scatac/"
scp -r "$source_path/database/sc_variant/table/scatac/difference_tf" "$username@$ip:$target_path/data/data/download/difference/"
scp -r "$source_path/database/sc_variant/table/scatac/difference_tf" "$username@$ip:$target_path/mysql/mysqlfile/scatac/"
scp -r "$source_path/database/sc_variant/table/scatac/sample_gene" "$username@$ip:$target_path/mysql/mysqlfile/scatac/"
scp -r "$source_path/database/sc_variant/table/scatac/sample_tf" "$username@$ip:$target_path/mysql/mysqlfile/scatac/"
scp -r $source_path/scATAC/**/**/**/*_difference_gene.h5ad $username@$ip:"$target_path/data/data/download/difference/difference_gene_h5ad/"
scp -r $source_path/scATAC/**/**/**/*_tf_activity_data.h5ad $username@$ip:"$target_path/data/data/download/difference/difference_tf_h5ad/"

# Difference gene enrichment
scp -r "$source_path/database/sc_variant/table/scatac/gene_enrichment_table_data.tar.gz" "$username@$ip:$target_path/data/data/download/enrichment/"
scp -r "$source_path/database/sc_variant/table/scatac/gene_enrichment_table" "$username@$ip:$target_path/data/data/download/enrichment/scatac/"
scp -r "$source_path/database/sc_variant/table/scatac/gene_enrichment" $username@$ip:"$target_path/data/data/download/enrichment/scatac/"

# variant
scp -r "$source_path/variant/finish/fine_mapping_hg19.tar.gz" "$username@$ip:$target_path/data/data/download/trait/"
scp -r "$source_path/variant/finish/fine_mapping_hg38.tar.gz" "$username@$ip:$target_path/data/data/download/trait/"
scp -r "$source_path/variant/finish/fine_mapping_trait.tar.gz" "$username@$ip:$target_path/data/data/download/trait/"

# Trait gene enrichment
scp -r "$source_path/database/sc_variant/table/magma/gene_enrichment_trait/gene_enrichment_trait_hg19.tar.gz" "$username@$ip:$target_path/data/data/download/enrichment/"
scp -r "$source_path/database/sc_variant/table/magma/gene_enrichment_trait/gene_enrichment_trait_hg38.tar.gz" "$username@$ip:$target_path/data/data/download/enrichment/"

scp -r "$source_path/database/sc_variant/table/magma/gene_enrichment_trait/hg19" "$username@$ip:$target_path/data/data/download/enrichment/trait/"
scp -r "$source_path/database/sc_variant/table/magma/gene_enrichment_trait/hg38" "$username@$ip:$target_path/data/data/download/enrichment/trait/"

# GENE
scp -r $source_path/gene/annotation/dbSNP/common_snp_chunk/* $username@$ip:"$target_path/mysql/mysqlfile/gene/annotation/dbSNP/common_snp_chunk/"
scp -r $source_path/gene/annotation/dbSUPER/*.txt $username@$ip:"$target_path/mysql/mysqlfile/gene/annotation/dbSUPER/"
scp -r $source_path/gene/annotation/GTEx/eqtl_chunk/* $username@$ip:"$target_path/mysql/mysqlfile/gene/annotation/GTEx/eqtl_chunk/"
scp -r $source_path/gene/annotation/GTEx/eqtl_chunk_chr_position/* $username@$ip:"$target_path/mysql/mysqlfile/gene/annotation/GTEx/eqtl_chunk_chr_position/"
scp -r $source_path/gene/annotation/gwasATLAS/gwasatlas_v20191115_risk_snp_*.txt $username@$ip:"$target_path/mysql/mysqlfile/gene/annotation/gwasATLAS/"
scp -r $source_path/gene/annotation/SEA/sea_v3_*.txt $username@$ip:"$target_path/mysql/mysqlfile/gene/annotation/SEA/"
scp -r $source_path/gene/annotation/SEdb/sedb_v2_super_enhancer_hg19.txt $username@$ip:"$target_path/mysql/mysqlfile/gene/annotation/SEdb/"
scp -r $source_path/gene/annotation/SEdb/sedb_v2_super_enhancer_hg38.txt $username@$ip:"$target_path/mysql/mysqlfile/gene/annotation/SEdb/"
scp -r $source_path/gene/annotation/SEdb/enhancer_chunk/* $username@$ip:"$target_path/mysql/mysqlfile/gene/annotation/SEdb/enhancer_chunk/"

scp -r $source_path/gene/annotation/dbSNP/dbsnp_common_snp_hg19.tar.gz $username@$ip:"$target_path/data/data/download/gene/hg19/"
scp -r $source_path/gene/annotation/dbSNP/dbsnp_common_snp_hg38.tar.gz $username@$ip:"$target_path/data/data/download/gene/hg38/"
scp -r $source_path/gene/annotation/dbSUPER/dbsuper_super_enhancer_hg19.tar.gz $username@$ip:"$target_path/data/data/download/gene/hg19/"
scp -r $source_path/gene/annotation/dbSUPER/dbsuper_super_enhancer_hg38.tar.gz $username@$ip:"$target_path/data/data/download/gene/hg38/"
scp -r $source_path/gene/annotation/GTEx/gtex_v10_eqtl_hg19.tar.gz $username@$ip:"$target_path/data/data/download/gene/hg19/"
scp -r $source_path/gene/annotation/GTEx/gtex_v10_eqtl_hg38.tar.gz $username@$ip:"$target_path/data/data/download/gene/hg38/"
scp -r $source_path/gene/annotation/gwasATLAS/gwasatlas_v20191115_risk_snp_hg19.tar.gz $username@$ip:"$target_path/data/data/download/gene/hg19/"
scp -r $source_path/gene/annotation/gwasATLAS/gwasatlas_v20191115_risk_snp_hg38.tar.gz $username@$ip:"$target_path/data/data/download/gene/hg38/"
scp -r $source_path/gene/annotation/SEA/sea_v3_enhancer_hg19.tar.gz $username@$ip:"$target_path/data/data/download/gene/hg19/"
scp -r $source_path/gene/annotation/SEA/sea_v3_enhancer_hg38.tar.gz $username@$ip:"$target_path/data/data/download/gene/hg38/"
scp -r $source_path/gene/annotation/SEA/sea_v3_super_enhancer_hg19.tar.gz $username@$ip:"$target_path/data/data/download/gene/hg19/"
scp -r $source_path/gene/annotation/SEA/sea_v3_super_enhancer_hg38.tar.gz $username@$ip:"$target_path/data/data/download/gene/hg38/"
scp -r $source_path/gene/annotation/SEdb/sedb_v2_enhancer_hg19.tar.gz $username@$ip:"$target_path/data/data/download/gene/hg19/"
scp -r $source_path/gene/annotation/SEdb/sedb_v2_enhancer_hg38.tar.gz $username@$ip:"$target_path/data/data/download/gene/hg38/"
scp -r $source_path/gene/annotation/SEdb/sedb_v2_super_enhancer_hg19.tar.gz $username@$ip:"$target_path/data/data/download/gene/hg19/"
scp -r $source_path/gene/annotation/SEdb/sedb_v2_super_enhancer_hg38.tar.gz $username@$ip:"$target_path/data/data/download/gene/hg38/"

```

```shell
tar -xzvf /mnt/data3/workspace/SCVdb/data/data/download/trait/fine_mapping_hg19.tar.gz -C /mnt/data3/workspace/SCVdb/data/data/download/variant/
tar -xzvf /mnt/data3/workspace/SCVdb/data/data/download/trait/fine_mapping_hg38.tar.gz -C /mnt/data3/workspace/SCVdb/data/data/download/variant/
tar -xzvf /mnt/data3/workspace/SCVdb/data/data/download/trait/fine_mapping_trait.tar.gz -C /mnt/data3/workspace/SCVdb/data/data/download/variant/

```
